// @goatui/components/dist/collection/components/navigation/header/header-brand/header-brand.js
import { Component, Fragment, h, Prop } from '@stencil/core';
export class HeaderBrand {
  constructor() {
    this.logo = '';
    this.name = '';
    this.href = '#';
    this.subTitle = '';
  }
  render() {
    const isLogoSVG = this.logo.endsWith('.svg');
    return (h("div", { class: 'header-brand' },
      h("goat-link", { class: 'brand-link full-height no-style', href: this.href },
        h("div", { class: 'brand' },
          (() => {
            if (this.logo) {
              if (isLogoSVG) {
                return h("goat-svg", { class: 'logo inherit', src: this.logo });
              }
              else {
                return h("img", { src: this.logo, class: 'logo', alt: this.name });
              }
            }
          })(),
          this.name && h("span", { class: 'brand-name' }, this.name))),
      (() => {
        if (this.subTitle)
          return h(Fragment, null,
            h("goat-divider", { vertical: true, class: 'subtitle-divider' }),
            h("div", { class: 'subtitle' }, this.subTitle));
      })()));
  }
  static get is() { return "goat-header-brand"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() { return {
    "$": ["header-brand.scss"]
  }; }
  static get styleUrls() { return {
    "$": ["header-brand.css"]
  }; }
  static get properties() { return {
    "logo": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "logo",
      "reflect": false,
      "defaultValue": "''"
    },
    "name": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "name",
      "reflect": false,
      "defaultValue": "''"
    },
    "href": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "href",
      "reflect": false,
      "defaultValue": "'#'"
    },
    "subTitle": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "sub-title",
      "reflect": false,
      "defaultValue": "''"
    }
  }; }
}
