// @goatui/components/dist/collection/components/navigation/tabs/tabs-list/tabs-list.js
import { Component, Element, h, Host, Listen, Prop } from '@stencil/core';
export class TabsList {
  constructor() {
    this.variant = 'line';
    this.managed = false;
  }
  tabClick(evt) {
    if (!this.managed) {
      this.deselectAllTabs();
      evt.target.selected = true;
      if (evt.detail.target) {
      }
      else {
        console.warn("goat-tabs:: No target associated");
      }
    }
  }
  deselectAllTabs() {
    const tabs = this.elm.querySelectorAll('goat-tab');
    tabs.forEach(tab => {
      tab.selected = false;
    });
  }
  render() {
    return (h(Host, null,
      h("div", { class: { 'tabs-list': true, [`variant-${this.variant}`]: true } },
        h("div", { class: "tabs-container" },
          h("slot", null)))));
  }
  static get is() { return "goat-tabs-list"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() { return {
    "$": ["tabs-list.scss"]
  }; }
  static get styleUrls() { return {
    "$": ["tabs-list.css"]
  }; }
  static get properties() { return {
    "variant": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "'line' | 'contained'",
        "resolved": "\"contained\" | \"line\"",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "variant",
      "reflect": false,
      "defaultValue": "'line'"
    },
    "managed": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "managed",
      "reflect": false,
      "defaultValue": "false"
    }
  }; }
  static get elementRef() { return "elm"; }
  static get listeners() { return [{
      "name": "goat:tab-click",
      "method": "tabClick",
      "target": undefined,
      "capture": false,
      "passive": false
    }]; }
}
