// @goatui/components/dist/components/goat-breadcrumb.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';

const breadcrumbCss = "*{box-sizing:border-box}:host{font-family:var(--font-family-base)}::selection{color:var(--color-white, #ffffff);background:var(--color-primary, var(--color-blue-60, #0f62fe))}::-webkit-scrollbar{width:1rem;height:1rem}::-webkit-scrollbar-track{background:#f1f1f1}::-webkit-scrollbar-thumb{background:#888}::-webkit-scrollbar-thumb :hover{background:#555}.sr-only{position:absolute;width:1px;height:1px;padding:0;margin:-1px;overflow:hidden;clip:rect(0, 0, 0, 0);border:0}.center-content{display:flex;align-items:center;justify-content:center}:host([hidden]){display:none}@keyframes reveal{0%{opacity:0;transform:translateY(8px)}}.breadcrumb{display:flex;align-items:center}slot::slotted(*)::before{display:inline-block;padding-right:0.5rem;padding-left:0.5rem;color:var(--border-color);content:\"/\"}slot::slotted(*:first-child)::before{display:none}";

const Breadcrumb = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
  }
  componentWillLoad() {
    this.elm.querySelectorAll('goat-breadcrumb-item').forEach((item, i) => {
      item.position = i + 1 + '';
    });
  }
  render() {
    return (h(Host, { itemscope: true, itemtype: 'http://schema.org/BreadcrumbList' }, h("div", { class: 'breadcrumb' }, h("slot", null))));
  }
  get elm() { return this; }
  static get style() { return breadcrumbCss; }
}, [1, "goat-breadcrumb"]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["goat-breadcrumb"];
  components.forEach(tagName => { switch (tagName) {
    case "goat-breadcrumb":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, Breadcrumb);
      }
      break;
  } });
}

const GoatBreadcrumb = Breadcrumb;
const defineCustomElement = defineCustomElement$1;

export { GoatBreadcrumb, defineCustomElement };
