// @goatui/components/dist/collection/components/content/avatar/avatar.js
import { Component, Element, h, Host, Prop } from '@stencil/core';
/**
 * @name Avatar
 * @description Avatars in their simplest form display content within a circular container.
 * @example <goat-avatar size="5rem" name="Shivaji Varma" src="/assets/img/avatar.png"></goat-avatar>
 */
export class Avatar {
  constructor() {
    /**
     * Avatar size.
     */
    this.size = '2rem';
    this.name = '';
    this.src = '';
  }
  getInitials() {
    const name = this.name.split(' ');
    let firstName = name[0] ? name[0].charAt(0).toUpperCase() : '';
    let lastName = name[1] ? name[1].charAt(0).toUpperCase() : '';
    return `${firstName}${lastName}`;
  }
  getFontSize() {
    const size = this.size;
    const fontSize = this.size.match(/^\d+(\.\d{1,2})?/)[0];
    // @ts-ignore
    return (fontSize * 4 / 10) + size.replace(/^\d+(\.\d{1,2})?/, '');
  }
  render() {
    const cssCls = ['avatar'];
    if (this.src) {
      cssCls.push('avatar-image');
    }
    else {
      cssCls.push('avatar-initials');
    }
    return (h(Host, { title: this.name },
      h("div", { class: 'avatar-container' },
        h("div", { class: cssCls.join(' '), style: { width: this.size, height: this.size, fontSize: this.getFontSize() } }, (() => {
          if (this.src) {
            return h("img", { class: 'image', src: this.src, alt: this.name });
          }
          else {
            return h("div", { class: 'initials' }, this.getInitials());
          }
        })()))));
  }
  static get is() { return "goat-avatar"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() { return {
    "$": ["avatar.scss"]
  }; }
  static get styleUrls() { return {
    "$": ["avatar.css"]
  }; }
  static get properties() { return {
    "size": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Avatar size."
      },
      "attribute": "size",
      "reflect": false,
      "defaultValue": "'2rem'"
    },
    "name": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "name",
      "reflect": false,
      "defaultValue": "''"
    },
    "src": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "src",
      "reflect": false,
      "defaultValue": "''"
    }
  }; }
  static get elementRef() { return "elm"; }
}
