// @goatui/components/dist/esm/goat-checkbox_3.entry.js
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './index-8d2c9e11.js';
import { g as getComponentIndex, d as debounceEvent, a as isMobile, b as isOutOfViewport } from './utils-3216c69d.js';

const checkboxCss = "*{box-sizing:border-box}:host{font-family:var(--font-family-base)}::selection{color:var(--color-white, #ffffff);background:var(--color-primary, var(--color-blue-60, #0f62fe))}::-webkit-scrollbar{width:1rem;height:1rem}::-webkit-scrollbar-track{background:#f1f1f1}::-webkit-scrollbar-thumb{background:#888}::-webkit-scrollbar-thumb :hover{background:#555}.sr-only{position:absolute;width:1px;height:1px;padding:0;margin:-1px;overflow:hidden;clip:rect(0, 0, 0, 0);border:0}.center-content{display:flex;align-items:center;justify-content:center}:host([hidden]){display:none}@keyframes reveal{0%{opacity:0;transform:translateY(8px)}}@keyframes checkbox-check{0%{height:0;width:0;border-color:var(--color-white, #ffffff);transform:translate3d(0, 0, 0) rotate(45deg)}33%{height:0;width:33%;transform:translate3d(0, 0, 0) rotate(45deg)}100%{height:67%;width:33%;border-color:var(--color-white, #ffffff);transform:translate3d(-50%, -40%, 0) rotate(45deg)}}@keyframes checkbox-slash{0%{width:0;border-color:var(--color-white, #ffffff)}33%{width:33%}100%{width:67%;border-color:var(--color-white, #ffffff)}}:host{display:inline-block;color:var(--text-primary, var(--color-gray-100, #161616))}.checkbox{position:relative;display:flex;align-items:center}.checkbox.has-content{gap:var(--spacing-2, 0.5rem)}.checkbox .input-native{position:absolute;width:1px;height:1px;padding:0;margin:-1px;overflow:hidden;clip:rect(0, 0, 0, 0);border:0}.checkbox .box{position:relative;background:var(--field-01, var(--color-gray-10, #f4f4f4));border:1px solid var(--ui-04, var(--color-gray-50, #8d8d8d));border-radius:var(--border-radius, 3px);cursor:pointer;display:flex;align-items:center;justify-content:center}.checkbox .box:before{pointer-events:none;content:\"\";display:block;width:100%;height:100%;box-sizing:border-box;border-radius:var(--border-radius, 3px);border:1px solid transparent}.checkbox.rounded .box{border-radius:50%}.checkbox.rounded .box:before{border-radius:50%}.checkbox .tick{position:absolute;width:0;height:0}.checkbox.has-focus:not(.active) .box{outline:2px solid var(--color-primary, var(--color-blue-60, #0f62fe));border-color:transparent}.checkbox.has-focus:not(.active):not(.state-checked) .box:before{border-color:var(--icon-01, var(--color-gray-100, #161616))}:host(.light) .box{background-color:var(--field-02, #ffffff)}.checkbox.state-checked .box{background:var(--color-primary, var(--color-blue-60, #0f62fe));border-color:transparent}.checkbox.state-checked .tick{width:33%;height:67%;border-right:3px solid transparent;border-bottom:3px solid transparent;transform:rotate(45deg);transform-origin:0 100%}.checkbox.state-checked .tick.animate{animation:checkbox-check 125ms 250ms cubic-bezier(0.4, 0, 0.23, 1) forwards}.checkbox.state-checked .tick:not(.animate){border-color:var(--color-white, #ffffff);top:-15%;left:15%}.checkbox.state-checked:hover .box{background:var(--color-primary-70, var(--color-blue-70, #0043ce))}.checkbox.state-checked.active .box{border-color:var(--color-white, #ffffff)}.checkbox.state-checked.has-focus:not(.active) .box{outline:2px solid var(--color-primary, var(--color-blue-60, #0f62fe));border-color:var(--color-white, #ffffff)}.checkbox.state-intermediate .box{background:var(--color-primary, var(--color-blue-60, #0f62fe));border-color:transparent}.checkbox.state-intermediate .tick{width:67%;height:0;margin:auto;border-right:3px solid transparent;border-bottom:3px solid transparent;transform-origin:0 100%}.checkbox.state-intermediate .tick.animate{animation:checkbox-slash 125ms 250ms cubic-bezier(0.4, 0, 0.23, 1) forwards}.checkbox.state-intermediate .tick:not(.animate){border-color:var(--color-white, #ffffff)}.checkbox.state-intermediate:hover .box{background:var(--color-primary-90, var(--color-blue-90, #001d6c))}.checkbox.state-intermediate.active .box{border-color:var(--color-white, #ffffff)}.checkbox.state-intermediate.has-focus:not(.active) .box{outline:2px solid var(--color-primary, var(--color-blue-60, #0f62fe));border-color:var(--color-white, #ffffff)}.checkbox.size-sm{font-size:var(--text-sm-font-size, 0.875rem);line-height:var(--text-sm-line-height, 1.25rem);letter-spacing:var(--text-sm-letter-spacing, 0rem)}.checkbox.size-sm .box{width:1.25rem;height:1.25rem}.checkbox.size-md{font-size:var(--text-md-font-size, 1rem);line-height:var(--text-md-line-height, 1.5rem);letter-spacing:var(--text-md-letter-spacing, 0rem)}.checkbox.size-md .box{width:1.5rem;height:1.5rem}.checkbox.size-lg{font-size:var(--text-lg-font-size, 1.125rem);line-height:var(--text-lg-line-height, 1.75rem);letter-spacing:var(--text-lg-letter-spacing, 0rem)}.checkbox.size-lg .box{width:1.75rem;height:1.75rem}:host-context(.input-state-success) .checkbox .box{border-color:var(--color-success, #198038)}:host-context(.input-state-warning) .checkbox .box{border-color:var(--color-warning, #8e6a00)}:host-context(.input-state-error) .checkbox .box{border-color:var(--color-error, #da1e28)}.checkbox.readonly{cursor:default;color:var(--text-disabled, var(--color-gray-30, #c6c6c6)) !important;border-color:var(--text-disabled, var(--color-gray-30, #c6c6c6)) !important}.checkbox.readonly.state-checked .box{background:var(--text-disabled, var(--color-gray-30, #c6c6c6)) !important}.checkbox.readonly .box{border:1px solid var(--text-disabled, var(--color-gray-30, #c6c6c6)) !important}.checkbox.readonly .tick{border-color:var(--disabled-01, var(--color-gray-10, #f4f4f4)) !important}.checkbox.disabled{cursor:not-allowed;color:var(--text-disabled, var(--color-gray-30, #c6c6c6)) !important;border-color:var(--text-disabled, var(--color-gray-30, #c6c6c6)) !important}.checkbox.disabled.state-checked .box{background:var(--text-disabled, var(--color-gray-30, #c6c6c6)) !important}.checkbox.disabled .box{border:1px solid var(--text-disabled, var(--color-gray-30, #c6c6c6)) !important}.checkbox.disabled .tick{border-color:var(--disabled-01, var(--color-gray-10, #f4f4f4)) !important}";

const Checkbox = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.goatChange = createEvent(this, "goat:change", 7);
    this.goatBlur = createEvent(this, "goat:blur", 7);
    this.goatFocus = createEvent(this, "goat:focus", 7);
    this.gid = getComponentIndex();
    /**
     * The input field name.
     */
    this.name = `goat-input-${this.gid}`;
    /**
     * The input field value.
     */
    this.value = false;
    this.intermediate = false;
    this.rounded = false;
    /**
     * The button size.
     * Possible values are: `"sm"`, `"md"`, `"lg"`. Defaults to `"md"`.
     */
    this.size = 'md';
    /**
     * If true, required icon is show. Defaults to `false`.
     */
    this.required = false;
    this.readonly = false;
    /**
     * If true, the user cannot interact with the button. Defaults to `false`.
     */
    this.disabled = false;
    this.configAria = {};
    this.mouseDownHandler = () => {
      this.isActive = true;
    };
    this.keyDownHandler = (evt) => {
      if (evt.key == ' ') {
        evt.preventDefault();
        this.isActive = true;
        this.clickHandler(evt);
      }
    };
    this.tabindex = 1;
    this.hasFocus = false;
    this.animate = true;
    this.isActive = false;
    this.slotHasContent = false;
    this.clickHandler = (ev) => {
      if (!this.disabled && !this.readonly) {
        this.value = !JSON.parse(this.nativeInput.value);
        this.intermediate = false;
        this.goatChange.emit(ev);
        this.iconContainer.focus();
      }
    };
    this.blurHandler = (ev) => {
      this.hasFocus = false;
      this.goatBlur.emit(ev);
    };
    this.focusHandler = (ev) => {
      this.hasFocus = true;
      this.goatFocus.emit(ev);
    };
  }
  async getComponentId() {
    return this.gid;
  }
  /**
   * Sets focus on the native `input` in `goat-input`. Use this method instead of the global
   * `input.focus()`.
   */
  async setFocus() {
    if (this.nativeInput) {
      this.nativeInput.focus();
    }
  }
  /**
   * Sets blur on the native `input` in `goat-input`. Use this method instead of the global
   * `input.blur()`.
   */
  async setBlur() {
    if (this.nativeInput) {
      this.nativeInput.blur();
    }
  }
  windowMouseUp() {
    if (this.isActive)
      this.isActive = false;
  }
  windowKeyUp(evt) {
    if (this.isActive && (evt.key == ' '))
      this.isActive = false;
  }
  componentWillLoad() {
    // If the ion-input has a tabindex attribute we get the value
    // and pass it down to the native input, then remove it from the
    // goat-input to avoid causing tabbing twice on the same element
    if (this.elm.hasAttribute('tabindex')) {
      const tabindex = this.elm.getAttribute('tabindex');
      this.tabindex = tabindex !== null ? tabindex : undefined;
      this.elm.removeAttribute('tabindex');
    }
    this.elm.getAttributeNames().forEach((name) => {
      if (name.includes('aria-')) {
        this.configAria[name] = this.elm.getAttribute(name);
        this.elm.removeAttribute(name);
      }
    });
    this.slotHasContent = this.elm.hasChildNodes();
  }
  render() {
    return (h(Host, { "has-focus": this.hasFocus, active: this.isActive }, h("label", { class: {
        'checkbox': true,
        'state-checked': this.value,
        'state-intermediate': !this.value && this.intermediate,
        [`size-${this.size}`]: true,
        'has-focus': this.hasFocus,
        'active': this.isActive,
        'disabled': this.disabled,
        'readonly': this.readonly,
        'required': this.required,
        'rounded': this.rounded,
        'has-content': this.slotHasContent,
      } }, h("div", Object.assign({ class: 'box', tabindex: this.tabindex, ref: elm => this.iconContainer = elm, onKeyUp: (evt) => {
        if (evt.keyCode === 13) {
          this.clickHandler(evt);
        }
      }, onMouseDown: this.mouseDownHandler, onKeyDown: this.keyDownHandler, onBlur: this.blurHandler, onFocus: this.focusHandler, role: "checkbox", "aria-disabled": this.disabled + '', "aria-required": this.required + '', "aria-checked": this.value + '' }, this.configAria), h("div", { class: {
        'tick': true,
        'animate': this.animate,
      } })), h("input", { type: 'checkbox', class: 'input-native', name: this.name, value: this.value + '', checked: this.value, "aria-hidden": 'true', required: this.required, ref: elm => this.nativeInput = elm, tabindex: '-1', onClick: this.clickHandler }), h("div", { class: 'slot-container' }, h("slot", null)))));
  }
  get elm() { return getElement(this); }
};
Checkbox.style = checkboxCss;

const formControlCss = "*{box-sizing:border-box}:host{font-family:var(--font-family-base)}::selection{color:var(--color-white, #ffffff);background:var(--color-primary, var(--color-blue-60, #0f62fe))}::-webkit-scrollbar{width:1rem;height:1rem}::-webkit-scrollbar-track{background:#f1f1f1}::-webkit-scrollbar-thumb{background:#888}::-webkit-scrollbar-thumb :hover{background:#555}.sr-only{position:absolute;width:1px;height:1px;padding:0;margin:-1px;overflow:hidden;clip:rect(0, 0, 0, 0);border:0}.center-content{display:flex;align-items:center;justify-content:center}:host([hidden]){display:none}@keyframes reveal{0%{opacity:0;transform:translateY(8px)}}:host{display:block;margin-bottom:var(--spacing-4, 1rem)}.form-control{display:flex;flex-direction:column}.form-control slot::slotted(goat-input),.form-control slot::slotted(goat-textarea),.form-control slot::slotted(goat-checkbox),.form-control slot::slotted(goat-radio),.form-control slot::slotted(goat-code-editor){margin-bottom:0.375rem}.label{font-size:var(--text-sm-font-size, 0.875rem);line-height:var(--text-sm-line-height, 1.25rem);letter-spacing:var(--text-sm-letter-spacing, 0rem);font-weight:var(--font-weight-medium, 500);margin-bottom:0.375rem;color:var(--text-secondary, var(--color-gray-70, #525252));overflow:hidden;white-space:nowrap;text-overflow:ellipsis}.label .required{color:var(--danger-02, #da1e28);padding-right:var(--spacing-1, 0.25rem)}.subtitle{font-size:var(--text-xs-font-size, 0.75rem);line-height:var(--text-xs-line-height, 1.125rem);letter-spacing:var(--text-xs-letter-spacing, 0rem);font-weight:var(--font-weight-regular, normal)}.subtitle .subtitle-caption{color:var(--text-05, var(--color-gray-60, #6f6f6f))}.subtitle .subtitle-error{color:var(--support-error-inverse, #fa4d56)}.subtitle .subtitle-success{color:var(--support-success-inverse, #42be65)}.subtitle .subtitle-warning{color:var(--support-warning-inverse, #f1c21b)}.form-control.inline{flex-direction:row;align-items:center}.form-control.inline .label{margin-bottom:0;margin-inline-end:var(--spacing-2, 0.5rem)}";

const FormControl = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.inline = false;
    /**
     * Whether the form control is required.
     */
    this.required = false;
  }
  componentDidLoad() {
    this.elm.setAttribute('role', 'group');
    for (const compName of ['goat-input', 'goat-textarea', 'goat-select', 'goat-checkbox', 'goat-radio', 'goat-code-editor']) {
      this.controlElm = this.elm.querySelector(`${compName}`);
      this.passRequiredToField(this.required);
      this.passLabelToField(this.label);
    }
  }
  passRequiredToField(required) {
    if (this.controlElm) {
      // @ts-ignore
      this.controlElm.required = required;
    }
  }
  passLabelToField(label) {
    if (this.controlElm) {
      // @ts-ignore
      const oldProps = this.controlElm.configAria;
      // @ts-ignore
      this.controlElm.configAria = Object.assign({ 'aria-label': label }, oldProps);
    }
  }
  componentShouldUpdate(newVal, _oldVal, propName) {
    if (propName === 'required') {
      this.passRequiredToField(newVal);
    }
    else if (propName === 'label') {
      this.passLabelToField(newVal);
    }
  }
  renderLabel() {
    if (this.label)
      return h("label", { class: 'label' }, this.required && h("span", { class: 'required' }, "*"), this.label);
  }
  renderSubtitle() {
    if (this.error)
      return h("div", { class: 'subtitle-error' }, this.error);
    else if (this.success)
      return h("div", { class: 'subtitle-success' }, this.success);
    else if (this.warning)
      return h("div", { class: 'subtitle-warning' }, this.warning);
    else if (this.caption)
      return h("div", { class: 'subtitle-caption' }, this.caption);
  }
  getInputState() {
    if (this.error)
      return 'input-state-error';
    else if (this.success)
      return 'input-state-success';
    else if (this.warning)
      return 'input-state-warning';
    return 'input-state-default';
  }
  render() {
    return (h(Host, { class: {
        [this.getInputState()]: true,
      } }, h("div", { class: { 'form-control': true, 'inline': this.inline } }, this.renderLabel(), h("div", { class: 'field' }, h("slot", null)), h("div", { class: 'subtitle' }, this.renderSubtitle()))));
  }
  get elm() { return getElement(this); }
};
FormControl.style = formControlCss;

const selectCss = "*{box-sizing:border-box}:host{font-family:var(--font-family-base)}::selection{color:var(--color-white, #ffffff);background:var(--color-primary, var(--color-blue-60, #0f62fe))}::-webkit-scrollbar{width:1rem;height:1rem}::-webkit-scrollbar-track{background:#f1f1f1}::-webkit-scrollbar-thumb{background:#888}::-webkit-scrollbar-thumb :hover{background:#555}.sr-only{position:absolute;width:1px;height:1px;padding:0;margin:-1px;overflow:hidden;clip:rect(0, 0, 0, 0);border:0}.center-content{display:flex;align-items:center;justify-content:center}:host([hidden]){display:none}@keyframes reveal{0%{opacity:0;transform:translateY(8px)}}:host{display:block;margin-bottom:var(--spacing-3, 0.75rem);--input-border-radius:var(--border-radius, 3px);--input-border-style:solid;color:var(--text-primary, var(--color-gray-100, #161616))}.input-container{position:relative;display:flex;align-items:center;justify-content:center;background-color:var(--field-01, var(--color-gray-10, #f4f4f4));border:1px var(--input-border-style) var(--ui-04, var(--color-gray-50, #8d8d8d));border-radius:var(--input-border-radius);overflow:hidden;cursor:text;font-weight:var(--font-weight-regular);gap:var(--spacing-2, 0.5rem)}.input-container .input{flex:1;border:none;outline:none;background:none;width:100%;cursor:inherit;padding:0;color:var(--text-primary, var(--color-gray-100, #161616))}.input-container .input::placeholder{color:var(--text-placeholder, var(--color-gray-40, #a8a8a8))}.input-container .slot-container{display:flex;align-items:center;justify-content:center;line-height:0}.input-container .slot-container slot::slotted(*){padding-bottom:0 !important;margin-bottom:0 !important;color:var(--color-neutral-400, )}.input-container .slot-container slot::slotted(*) goat-icon{--color:var(--color-neutral-400, )}.input-container.start-slot-has-content .input{padding-left:var(--spacing-2, 0.5rem)}.input-container.end-slot-has-content .input{padding-right:var(--spacing-2, 0.5rem)}.input-container:not(.start-slot-has-content) .slot-container.start{display:none}.input-container:not(.end-slot-has-content) .slot-container.end{display:none}.input-container .input-action{cursor:pointer}.input-container .input-action:hover{--text-color:var(--text-secondary)}:host(.light) .input-container{background-color:var(--field-02, #ffffff)}.input-container.has-focus{outline:2px solid var(--color-primary, var(--color-blue-60, #0f62fe));border-color:transparent}:host-context(.input-statesuccess) .input-container{border-color:var(--support-success, #198038)}:host-context(.input-statesuccess) .input-container.has-focus{outline:2px solid var(--support-success, #198038)}:host-context(.input-statewarning) .input-container{border-color:var(--support-warning, #f1c21b)}:host-context(.input-statewarning) .input-container.has-focus{outline:2px solid var(--support-warning, #f1c21b)}:host-context(.input-stateerror) .input-container{border-color:var(--support-error, #da1e28)}:host-context(.input-stateerror) .input-container.has-focus{outline:2px solid var(--support-error, #da1e28)}:host([size=sm]) .input-container{padding:0.5rem 0.75rem;min-height:2.375rem}:host([size=sm]) .input-container .input{font-size:var(--text-sm-font-size, 0.875rem);line-height:var(--text-sm-line-height, 1.25rem);letter-spacing:var(--text-sm-letter-spacing, 0rem)}:host([size=sm]) .input-container slot::slotted(*){font-size:var(--text-sm-font-size, 0.875rem);line-height:var(--text-sm-line-height, 1.25rem);letter-spacing:var(--text-sm-letter-spacing, 0rem)}:host([size=md]) .input-container{padding:0.625rem 0.875rem;min-height:2.875rem}:host([size=md]) .input-container .input{font-size:var(--text-md-font-size, 1rem);line-height:var(--text-md-line-height, 1.5rem);letter-spacing:var(--text-md-letter-spacing, 0rem)}:host([size=md]) .input-container slot::slotted(*){font-size:var(--text-md-font-size, 1rem);line-height:var(--text-md-line-height, 1.5rem);letter-spacing:var(--text-md-letter-spacing, 0rem)}:host([size=lg]) .input-container{padding:0.75rem 1rem;min-height:3.375rem}:host([size=lg]) .input-container .input{font-size:var(--text-lg-font-size, 1.125rem);line-height:var(--text-lg-line-height, 1.75rem);letter-spacing:var(--text-lg-letter-spacing, 0rem)}:host([size=lg]) .input-container slot::slotted(*){font-size:var(--text-lg-font-size, 1.125rem);line-height:var(--text-lg-line-height, 1.75rem);letter-spacing:var(--text-lg-letter-spacing, 0rem)}:host([success]) .input-container{border-color:var(--color-success, #198038)}:host([error]) .input-container{border-color:var(--color-error, #da1e28);box-shadow:0 0 0 0.25rem rgba(239, 68, 68, 0.25)}:host([readonly]) .input-container{cursor:initial;background-color:var(--disabled-01, var(--color-gray-10, #f4f4f4)) !important;border-color:var(--disabled-01, var(--color-gray-10, #f4f4f4)) !important;opacity:1 !important}:host([disabled]){color:var(--text-disabled, var(--color-gray-30, #c6c6c6)) !important}:host([disabled]) .input-container{cursor:not-allowed;background-color:var(--disabled-01, var(--color-gray-10, #f4f4f4)) !important;border-color:var(--disabled-01, var(--color-gray-10, #f4f4f4)) !important;opacity:1 !important}:host([disabled]) .input-container .input{color:var(--text-disabled, var(--color-gray-30, #c6c6c6)) !important}input::-webkit-outer-spin-button,input::-webkit-inner-spin-button{-webkit-appearance:none;margin:0}input[type=number]{-moz-appearance:textfield}.dropdown{position:relative}.dropdown .dropdown-content{z-index:var(--z-index-dropdown-content, 40);position:absolute;width:max-content;line-height:0;transform:scale(0);transition:transform 0.1s ease-out 0s}.dropdown .chevron-down{transition:transform 0.1s ease-out 0s}.dropdown.is-open .dropdown-content{transform:scale(1)}.dropdown.is-open .chevron-down{transform:rotate(180deg)}.dropdown.bottom-right .dropdown-content{top:calc(100% + var(--spacing-2, 0.5rem));left:0;transform-origin:top}.dropdown.bottom-left .dropdown-content{top:calc(100% + var(--spacing-2, 0.5rem));right:0;transform-origin:top}.dropdown.top-right .dropdown-content{bottom:calc(100% + var(--spacing-2, 0.5rem));left:0;transform-origin:bottom}.dropdown.top-left .dropdown-content{bottom:calc(100% + var(--spacing-2, 0.5rem));right:0;transform-origin:bottom}.dropdown.center .dropdown-content{top:0;left:0;position:fixed;transform-origin:center;display:flex;align-items:center;width:100vw;height:100vh;justify-content:center;background:rgba(0, 0, 0, 0.5);pointer-events:none}.select .dropdown-content{width:100%;min-width:10rem}.select .menu{width:100%}.select .display-value{overflow:hidden;white-space:nowrap;text-overflow:ellipsis;cursor:pointer}.select .multi-select-value{padding-inline-end:var(--spacing-2, 0.5rem)}.select .multi-select-value:last-child{padding-inline-end:0}.select .input-container:not(.has-value) .display-value{color:var(--text-placeholder, var(--color-gray-40, #a8a8a8))}.select .start-search{height:10rem;display:flex;align-items:center;justify-content:center;flex-direction:column;gap:var(--spacing-3, 0.75rem)}";

const Select = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.goatChange = createEvent(this, "goat:change", 7);
    this.p4ActionClick = createEvent(this, "goat:action-click", 7);
    this.goatSearch = createEvent(this, "goat:search", 7);
    this.gid = getComponentIndex();
    /**
     * The input field name.
     */
    this.name = `goat-input-${this.gid}`;
    /**
     * The input field value.
     */
    this.value = '';
    this.multiple = false;
    /**
     * The select input size.
     * Possible values are: `"sm"`, `"md"`, `"lg"`. Defaults to `"md"`.
     */
    this.size = 'md';
    /**
     * Search type
     * Possible values are `"none"`, `"initial"`, `"contains"`, `"managed"`. Defaults to `"none"`.
     */
    this.search = 'none';
    /**
     * The input state.
     * Possible values are: `"success"`, `"error"`, `"warning"`, 'default'. Defaults to `"default"`.
     */
    this.state = 'default';
    /**
     * If true, required icon is show. Defaults to `false`.
     */
    this.required = false;
    /**
     * If true, the user cannot interact with the button. Defaults to `false`.
     */
    this.disabled = false;
    /**
     * If true, the user cannot interact with the button. Defaults to `false`.
     */
    this.readonly = false;
    this.showLoader = false;
    this.isOpen = false;
    this.configAria = {};
    /**
     *  [{
     *    label: 'Shivaji Varma',
     *    value: 'shivaji-varma'
     *  }]
     */
    this.items = [];
    this.positions = 'bottom-right,top-right,bottom-left,top-left';
    /**
     * If `true`, a clear icon will appear in the input when there is a value. Clicking it clears the input.
     */
    this.clearable = false;
    /**
     * Set the amount of time, in milliseconds, to wait to trigger the `goatChange` event after each keystroke.
     */
    this.debounce = 300;
    this.hasFocus = false;
    this.searchString = '';
    this.startSlotHasContent = false;
    this.endSlotHasContent = false;
    this.selectHandler = (selectItemValue) => {
      if (!this.disabled && !this.readonly) {
        this.addItem(selectItemValue);
      }
      this.closeList();
    };
    this.clearInput = () => {
      if (!this.disabled && !this.readonly) {
        this.removeItem(this.value);
      }
    };
    this.blurHandler = () => {
      this.hasFocus = false;
    };
    this.focusHandler = () => {
      this.hasFocus = true;
    };
    this.closeList = () => {
      if (!this.disabled && !this.readonly && this.isOpen) {
        this.isOpen = false;
        setTimeout(() => this.setFocus(), 100);
      }
    };
    this.openList = () => {
      if (!this.disabled && !this.readonly && !this.isOpen) {
        this.isOpen = true;
        if (this.search !== 'none') {
          this.searchString = '';
          setTimeout(() => {
            const dropdownContent = this.dropdownContentElm;
            this.dropdownContentHeight = dropdownContent.getBoundingClientRect().height;
            this.dropdownContentWidth = dropdownContent.getBoundingClientRect().width;
            this.fixPosition();
            this.nativeInput.focus();
          }, 100);
        }
        else {
          setTimeout(() => {
            const dropdownContent = this.dropdownContentElm;
            this.dropdownContentHeight = dropdownContent.getBoundingClientRect().height;
            this.dropdownContentWidth = dropdownContent.getBoundingClientRect().width;
            this.fixPosition();
          }, 100);
        }
      }
    };
    this.toggleList = () => {
      if (this.isOpen)
        this.closeList();
      else
        this.openList();
    };
    this.keyDownHandler = (evt) => {
      if (evt.key === 'Enter') {
        evt.preventDefault();
        this.toggleList();
      }
      else if (evt.key === 'ArrowDown') {
        if (this.isOpen) {
          console.log('inside select');
          evt.preventDefault();
          this.menuElm.setFocus();
        }
      }
      else if (evt.key === 'ArrowUp') {
        if (this.isOpen) {
          evt.preventDefault();
          this.menuElm.setFocus(); // focus on previous item
        }
      }
    };
    this.onInput = (ev) => {
      const input = ev.target;
      this.searchString = input.value || '';
      this.goatSearch.emit({ value: this.searchString });
    };
  }
  /**
   * Sets focus on the native `input` in `ion-input`. Use this method instead of the global
   * `input.focus()`.
   */
  async setFocus() {
    this.displayElement.focus();
  }
  /**
   * Sets blur on the native `input` in `goat-input`. Use this method instead of the global
   * `input.blur()`.
   */
  async setBlur() {
    if (this.nativeInput) {
      this.nativeInput.blur();
    }
  }
  debounceChanged() {
    this.goatSearch = debounceEvent(this.goatSearch, this.debounce);
  }
  windowClick(evt) {
    const path = (evt.path || evt.composedPath());
    for (const elm of path) {
      if (elm == this.elm)
        return;
    }
    this.isOpen = false;
  }
  menuItemClick(evt) {
    this.selectHandler(evt.detail.value);
  }
  async getComponentId() {
    return this.gid;
  }
  tagDismissClick(evt) {
    this.removeItem(evt.detail.value);
  }
  getValues() {
    if (this.value)
      return this.value.toString().split(',');
    else
      return [];
  }
  addItem(selectItemValue) {
    let value = this.getValues();
    if (!value.includes(selectItemValue)) {
      if (!this.multiple)
        value = [];
      value.push(selectItemValue);
      this.value = value.join(',');
      this.goatChange.emit({ value: this.value, newItem: this.getItemByValue(selectItemValue) });
    }
  }
  removeItem(selectItemValue) {
    let value = this.getValues();
    if (value.includes(selectItemValue)) {
      value = value.filter(item => item !== selectItemValue);
      this.value = value.join(',');
      this.goatChange.emit({ value: this.value, removedItem: this.getItemByValue(selectItemValue) });
    }
  }
  hasValue() {
    return (this.value || '').toString().length > 0;
  }
  getItemByValue(value) {
    if (this.items) {
      return this.items.find((item) => {
        return item.value == value;
      });
    }
  }
  getDisplayValue() {
    if (!this.multiple) {
      if (this.items) {
        const item = this.getItemByValue(this.value);
        if (item) {
          return item.label;
        }
      }
      if (!this.disabled && !this.readonly) {
        return this.placeholder;
      }
      else {
        return h("span", null, "\u00A0");
      }
    }
    else {
      if (!this.value && !this.disabled && !this.readonly) {
        return this.placeholder;
      }
      else {
        return h("span", null, "\u00A0");
      }
    }
  }
  componentWillLoad() {
    if (this.positions)
      this.position = this.positions.split(',')[0];
    this.elm.getAttributeNames().forEach((name) => {
      if (name.includes('aria-')) {
        this.configAria[name] = this.elm.getAttribute(name);
        this.elm.removeAttribute(name);
      }
    });
    this.startSlotHasContent = !!this.elm.querySelector('[slot="start"]');
    this.endSlotHasContent = !!this.elm.querySelector('[slot="end"]');
  }
  fixPosition() {
    if (this.isOpen && this.dropdownContentHeight && this.dropdownContentWidth) {
      if (isMobile()) {
        this.position = 'center';
        return;
      }
      const positions = this.positions.split(',');
      for (let i = 0; i < positions.length; i++) {
        const dropdownButtonRect = this.elm.getBoundingClientRect();
        const dropdownContentRect = {};
        if (positions[i] === 'bottom-right') {
          dropdownContentRect.top = dropdownButtonRect.top + dropdownButtonRect.height;
          dropdownContentRect.bottom = dropdownContentRect.top + this.dropdownContentHeight;
          dropdownContentRect.left = dropdownButtonRect.left;
          dropdownContentRect.right = dropdownButtonRect.left + this.dropdownContentWidth;
        }
        else if (positions[i] === 'top-right') {
          dropdownContentRect.top = dropdownButtonRect.top - this.dropdownContentHeight;
          dropdownContentRect.bottom = dropdownButtonRect.top;
          dropdownContentRect.left = dropdownButtonRect.left;
          dropdownContentRect.right = dropdownButtonRect.left + this.dropdownContentWidth;
        }
        else if (positions[i] === 'bottom-left') {
          dropdownContentRect.top = dropdownButtonRect.top + dropdownButtonRect.height;
          dropdownContentRect.bottom = dropdownContentRect.top + this.dropdownContentHeight;
          dropdownContentRect.left = dropdownButtonRect.left - this.dropdownContentWidth;
          dropdownContentRect.right = dropdownButtonRect.left;
        }
        else if (positions[i] === 'top-left') {
          dropdownContentRect.top = dropdownButtonRect.top - this.dropdownContentHeight;
          dropdownContentRect.bottom = dropdownButtonRect.top;
          dropdownContentRect.left = dropdownButtonRect.left - this.dropdownContentWidth;
          dropdownContentRect.right = dropdownButtonRect.left;
        }
        const isOut = isOutOfViewport(dropdownContentRect);
        if (!isOut.any) {
          this.position = positions[i];
          break;
        }
      }
    }
  }
  ;
  connectedCallback() {
    this.debounceChanged();
  }
  renderMultiSelectValues() {
    const values = this.getValues();
    if (this.multiple && values.length) {
      return values.map((value) => {
        const item = this.getItemByValue(value);
        if (item) {
          return (h("goat-tag", { filter: true, class: 'multi-select-value', value: item.value }, item.label));
        }
      });
    }
  }
  render() {
    return (h(Host, { "has-value": this.hasValue(), "has-focus": this.hasFocus, "is-open": this.isOpen, position: this.position }, h("div", { class: { 'dropdown': true, 'select': true, [this.position]: true, 'is-open': this.isOpen } }, h("div", { class: {
        'input-container': true,
        [`search-${this.search}`]: true,
        'has-focus': this.hasFocus,
        'disabled': this.disabled,
        'readonly': this.readonly,
        'has-value': this.hasValue(),
        'start-slot-has-content': this.startSlotHasContent,
        'end-slot-has-content': this.endSlotHasContent,
      } }, h("div", { class: 'slot-container start' }, h("slot", { name: 'start' })), h("div", { class: 'multi-select-values' }, this.renderMultiSelectValues()), (() => {
      if (this.search !== 'none' && this.isOpen) {
        return h("input", Object.assign({ class: 'input input-native', ref: input => this.nativeInput = input, type: 'text', name: this.name, value: this.searchString, placeholder: this.placeholder, onBlur: this.blurHandler, onFocus: this.focusHandler, onInput: this.onInput, onKeyDown: this.keyDownHandler }, this.configAria));
      }
      else {
        return h("div", Object.assign({ class: 'input display-value', tabindex: '0', ref: input => this.displayElement = input, "aria-disabled": this.disabled ? 'true' : null, onFocus: this.focusHandler, onBlur: this.blurHandler, onKeyDown: this.keyDownHandler, onClick: (evt) => {
            evt.preventDefault();
            this.toggleList();
          } }, this.configAria), this.getDisplayValue());
      }
    })(), this.clearable && !this.multiple && this.hasValue() &&
      h("goat-icon", { class: 'clear input-action', name: 'x-circle-fill', size: this.size, onClick: this.clearInput, role: 'button' }), h("div", { class: 'slot-container end' }, h("slot", { name: 'end' })), this.getModeIcon()), h("div", { class: 'dropdown-content', ref: (elm) => this.dropdownContentElm = elm }, this.isOpen && this.renderDropdownList()))));
  }
  getModeIcon() {
    if (this.showLoader) {
      return h("goat-spinner", { class: 'input-action rainbow' });
    }
    if (!this.disabled && !this.readonly)
      return h("goat-icon", { name: 'chevron-down', size: this.size, class: 'input-action chevron-down', role: 'button', onClick: this.toggleList });
  }
  renderDropdownList() {
    if (this.search === 'managed' && !this.items.length) {
      return h("goat-menu", { class: 'menu', ref: (el) => this.menuElm = el }, h("div", { class: 'start-search' }, h("goat-icon", { name: 'search', size: this.size }), h("goat-text", { shade: 'secondary' }, "Start typing to perform search")));
    }
    if (this.items) {
      const filteredItems = this.filterItems();
      return h("goat-menu", { class: 'menu', empty: filteredItems.length == 0, ref: (el) => this.menuElm = el }, (() => {
        return filteredItems.map((item) => {
          return h("goat-menu-item", { value: item.value }, item.label || item.value);
        });
      })());
    }
  }
  filterItems() {
    if (this.search === 'managed')
      return this.items;
    return this.items.filter((item) => {
      return (!this.searchString || item.label.toLocaleLowerCase().includes(this.searchString.toLocaleLowerCase()));
    });
  }
  get elm() { return getElement(this); }
  static get watchers() { return {
    "debounce": ["debounceChanged"]
  }; }
};
Select.style = selectCss;

export { Checkbox as goat_checkbox, FormControl as goat_form_control, Select as goat_select };
