// @goatui/components/dist/cjs/goat-link.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-7441634b.js');
const utils = require('./utils-652ef10c.js');

const linkCss = "*{box-sizing:border-box}:host{font-family:var(--font-family-base)}::selection{color:var(--color-white, #ffffff);background:var(--color-primary, var(--color-blue-60, #0f62fe))}::-webkit-scrollbar{width:1rem;height:1rem}::-webkit-scrollbar-track{background:#f1f1f1}::-webkit-scrollbar-thumb{background:#888}::-webkit-scrollbar-thumb :hover{background:#555}.sr-only{position:absolute;width:1px;height:1px;padding:0;margin:-1px;overflow:hidden;clip:rect(0, 0, 0, 0);border:0}.center-content{display:flex;align-items:center;justify-content:center}:host([hidden]){display:none}@keyframes reveal{0%{opacity:0;transform:translateY(8px)}}:host{display:inline-block}.link{border-radius:var(--border-radius, 3px);display:flex;align-items:center;flex-direction:row;cursor:pointer;text-decoration:var(--link-decoration, none);color:var(--link-color, var(--link-01, var(--color-blue-60, #0f62fe)))}.link:hover{color:var(--link-color-hover, var(--hover-primary-text, var(--color-blue-70, #0043ce)));text-decoration:var(--link-decoration-hover, underline)}.link:visited{color:var(--link-color-visited, var(--visited-link, #8a3ffc))}.link.active{color:var(--link-color-active, var(--text-primary, var(--color-gray-100, #161616)))}:host(.no-style) .link{color:inherit !important;text-decoration:none !important}:host(.no-decoration) .link{text-decoration:none !important}:host(.inline) .link{text-decoration:var(--link-decoration, underline)}.link.has-focus:not(.active){text-decoration:none;outline:2px solid var(--color-primary, var(--color-blue-60, #0f62fe))}";

const Link = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.gid = utils.getComponentIndex();
    this.hasFocus = false;
    this.isActive = false;
    this.tabindex = 1;
    this.blurHandler = () => {
      this.hasFocus = false;
    };
    this.focusHandler = () => {
      this.hasFocus = true;
    };
    this.mouseDownHandler = () => {
      this.isActive = true;
    };
    this.keyDownHandler = (evt) => {
      if (evt.key == 'Enter' || evt.key == ' ') {
        this.isActive = true;
      }
    };
  }
  windowMouseUp() {
    if (this.isActive)
      this.isActive = false;
  }
  windowKeyUp(evt) {
    if (this.isActive && (evt.key == 'Enter' || evt.key == ' '))
      this.isActive = false;
  }
  async triggerClick() {
    if (this.nativeInput) {
      this.nativeInput.click();
    }
  }
  componentWillLoad() {
    // If the ion-input has a tabindex attribute we get the value
    // and pass it down to the native input, then remove it from the
    // goat-input to avoid causing tabbing twice on the same element
    if (this.elm.hasAttribute('tabindex')) {
      const tabindex = this.elm.getAttribute('tabindex');
      this.tabindex = tabindex !== null ? tabindex : undefined;
      this.elm.removeAttribute('tabindex');
    }
    if (this.elm.hasAttribute('itemprop')) {
      const tabindex = this.elm.getAttribute('itemprop');
      this.itemprop = tabindex !== null ? tabindex : undefined;
      this.elm.removeAttribute('itemprop');
    }
  }
  render() {
    return (index.h(index.Host, { "has-focus": this.hasFocus, active: this.isActive }, index.h("a", { class: {
        'link': true,
        'has-focus': this.hasFocus,
        'active': this.isActive,
      }, href: this.href, target: this.target, itemprop: this.itemprop, ref: input => this.nativeInput = input, tabindex: this.tabindex, onBlur: this.blurHandler, onFocus: this.focusHandler, onMouseDown: this.mouseDownHandler, onKeyDown: this.keyDownHandler }, index.h("slot", null))));
  }
  get elm() { return index.getElement(this); }
};
Link.style = linkCss;

exports.goat_link = Link;
